x-pack/plugins/apm/server/lib/services/get_service_transaction_groups_metrics/index.ts

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Coordinate } from '../../../../typings/timeseries';
+import { EVENT_OUTCOME } from '../../../../common/elasticsearch_fieldnames';
+import { LatencyAggregationType } from '../../../../common/latency_aggregation_types';
+import { getLatencyValue } from '../../helpers/latency_aggregation_type';
+import { Setup, SetupTimeRange } from '../../helpers/setup_request';
+import { getTimeseriesDataForTransactionGroups } from '../get_service_transaction_groups/get_timeseries_data_for_transaction_groups';
+
+export async function getServiceTransactionGroupsMetrics({
+  serviceName,
+  transactionNames,
+  setup,
+  numBuckets,
+  searchAggregatedTransactions,
+  transactionType,
+  latencyAggregationType,
+}: {
+  serviceName: string;
+  transactionNames: string[];
+  setup: Setup & SetupTimeRange;
+  numBuckets: number;
+  searchAggregatedTransactions: boolean;
+  transactionType: string;
+  latencyAggregationType: LatencyAggregationType;
+}): Promise<
+  Record<
+    string,
+    {
+      latency: { timeseries: Coordinate[] };
+      throughput: { timeseries: Coordinate[] };
+      errorRate: { timeseries: Coordinate[] };
+    }
+  >
+> {
+  const { apmEventClient, start, end, esFilter } = setup;
+
+  const buckets = await getTimeseriesDataForTransactionGroups({
+    apmEventClient,
+    start,
+    end,
+    esFilter,
+    numBuckets,
+    searchAggregatedTransactions,
+    serviceName,
+    transactionNames,
+    transactionType,
+    latencyAggregationType,
+  });
+  const deltaAsMinutes = (end - start) / 1000 / 60;
+
+  return buckets.reduce((bucketAcc, bucket) => {
```

Comment:
We might be able to simplify this, e.g. by not using `.reduce()`.

Reply:
@dgieselaar I simplified it a bit, removed one of the reduce, take a look and let me know what you think: https://github.com/elastic/kibana/pull/88946/files#diff-d61f31a22963c65b64a5229d0110313821b1fafb87fd6fd0c2e3802e1dc799a7R118-R153

Reply:
Still uses .reduce() AFAICT? I was more thinking of something like:

```ts
{
	...,
 	latency:
		bucket.timeseries.buckets.map(dateBucket => ({
			x: dateBucket.key,
			y: getLatencyValue(dateBucket.latency)
		}))
}
```

Reply:
> Still uses .reduce() AFAICT?

I left only one reducer, with is going to return a `Record<string, {latency, throughput, errorRate}>`. 

Why do you think doing three loops, one for each metrics (latency, throughput, errorRate), is better than one like now?

PS I'm not saying mine is better than what you've suggested 😄

Reply:
I think three `.map()`s is fine. It's maybe 100-200 iterations extra, it's negligible.

Reply:
That is, assuming that the request to get comparison metrics only covers transaction groups on the current page. Otherwise it can be a lot more (500x vs 10x)

Reply:
Which reminds me, you probably have to use both request id + transaction names (stringified) as an invalidator